### PR TITLE
Go: Deal better with a single go.mod file which is not in the project root

### PR DESCRIPTION
**Pull request by owen-mc**
This PR relates to https://github.com/github/codeql-go-team/issues/303.

Still to do:
- [x] Add integration tests for detecting a bazel BUILD file and when there are go.mod files but also .go files which aren't underneath any of the go.mod
- [x] Tidy up the code further
- [x] Test on real world examples
- [x] QA test?

### CODE REVIEW

File: go/extractor/cli/go-autobuilder/go-autobuilder.go [hunk 1]
```diff
@@ -240,28 +241,109 @@ func getSourceDir() string {
...
+	sort.Strings(dirs)
+	for _, dir := range dirs {
+		if !strings.HasPrefix(dir, dirs[0]) {
+			return false
+		}
+	}
+	return true
+}
+
+func findGoModFiles(emitDiagnostics bool) (string, bool) {
```
Comment (smowton) [inline go/extractor/cli/go-autobuilder/go-autobuilder.go:265]: Add comment saying what the return elements mean

**Review by smowton — COMMENTED** [review]
Looks good to me!

**Review by smowton — DISMISSED** [review]
Changes look good; I don't completely follow the back-and-forth regarding what to do when there are go.mod files nested under a single unique root -- I guess the desired behaviour is that if we have a/go.mod, a/file1.go, a/b/go.mod and a/b/file2.go then we'll extract starting from a/, mirroring what we would do today if all those files had their a/ prefix removed? If so, looks good -- let's get a QA test done to check for hidden snags / the expected newly-working repos.

**Comment by owen-mc** [comment]
Actually what I've made it do in the example you gave is the same as it does now, i.e. fail to find the go.mod. We can improve it in future when we have more time to test.

**Comment by owen-mc** [comment]
I manually verified that this was improving our autobuilding on the following repos: https://github.com/tgbot-collection/YYeTsBot. I tried to find other repos but most either have a go.mod file in the root, multiple go.mod files or some .go files outside of the folder with the go.mod.

**Review by smowton — APPROVED** [review]

**Pull request merged**